crates/miden-tx/src/executor/mod.rs

```diff
@@ -222,6 +225,106 @@ impl TransactionExecutor {
 
         Ok(*stack_outputs)
     }
+
+    // CHECK CONSUMABILITY
+    // ============================================================================================
+
+    /// Executes the transaction with specified notes, returning the [NoteAccountExecution::Success]
+    /// if all notes has been consumed successfully and [NoteAccountExecution::Failure] if some note
+    /// returned an error.
+    ///
+    /// # Errors:
+    /// Returns an error if:
+    /// - If required data can not be fetched from the [DataStore].
+    /// - If the transaction host can not be created from the provided values.
+    /// - If the execution of the provided program fails on the stage other than note execution.
+    #[maybe_async]
+    pub(crate) fn try_execute_notes(
+        &self,
+        account_id: AccountId,
+        block_ref: BlockNumber,
+        notes: InputNotes<InputNote>,
+        tx_args: TransactionArgs,
+    ) -> Result<NoteAccountExecution, TransactionExecutorError> {
+        // Validate that notes were not created after the reference, and build the set of required
+        // block numbers
+        let mut ref_blocks: BTreeSet<BlockNumber> = BTreeSet::new();
+        for note in &notes {
+            if let Some(location) = note.location() {
+                if location.block_num() > block_ref {
+                    return Err(TransactionExecutorError::NoteBlockPastReferenceBlock(
+                        note.id(),
+                        block_ref,
+                    ));
+                }
+                ref_blocks.insert(location.block_num());
+            }
+        }
```

Comment:
This looks very similar to code on lines 116 - 127 above. Maybe makes sense to create a helper function for this?